fraud/service.go

```diff
@@ -0,0 +1,171 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/celestiaorg/celestia-node/header"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+var log = logging.Logger("fraud")
+
+// proofUnmarshaler aliases a function that parses data to `Proof`
+type proofUnmarshaler func([]byte) (Proof, error)
+
+// headerFetcher aliases a function that is used to fetch ExtendedHeader from store
+type headerFetcher func(context.Context, uint64) (*header.ExtendedHeader, error)
+
+// Validator aliases a function that validate pubsub incoming messages
+type Validator func(ctx context.Context, proofType ProofType, data []byte) pubsub.ValidationResult
+
+// service is propagating and validating Fraud Proofs
+// service implements Service interface.
+type service struct {
+	pubsub       *pubsub.PubSub
+	topics       map[ProofType]*pubsub.Topic
+	unmarshalers map[ProofType]proofUnmarshaler
+	getter       headerFetcher
+	mu           *sync.Mutex
+}
+
+func NewService(p *pubsub.PubSub, hstore header.Store) Service {
+	return &service{
+		pubsub: p, topics: make(map[ProofType]*pubsub.Topic),
+		unmarshalers: make(map[ProofType]proofUnmarshaler),
+		getter:       hstore.GetByHeight,
+		mu:           &sync.Mutex{},
+	}
+}
+
+func (f *service) Start(context.Context) error {
+	log.Info("fraud service is starting...")
+
+	// Add validator and default Unmarshaler for BEFP
+	err := f.AddValidator(BadEncoding, f.processIncoming)
+	if err != nil {
+		return err
+	}
+	err = f.RegisterUnmarshaler(BadEncoding, UnmarshalBEFP)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *service) Stop(context.Context) error {
+	log.Info("fraud service is stopping...")
+	return nil
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.Lock()
+	defer f.mu.Unlock()
```

Comment:
The function call here may take some time because of `pubsub.Join`. Let's not unlock on defer but once we are done reading unmarshallers.

Reply:
This read here is also a reason to use RWMutex instead

Reply:
topics map should have its own mutex.

Reply:
done